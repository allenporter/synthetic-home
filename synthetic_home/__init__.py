"""
.. include:: ../README.md
"""

__all__ = [
    "synthetic_home",
    "device_types",
    "exceptions",
    "registry",
    "tool",
]
