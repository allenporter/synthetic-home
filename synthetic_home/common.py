"""Common data model objects used across this library.

These are typically small objects reusable in different contexts.
"""

from dataclasses import dataclass

__all__ = ["DeviceInfo"]


@dataclass
class DeviceInfo:
    """Device model information."""

    model: str | None = None
    """The model name of the device e.g. 'Learning Thermostat'."""

    manufacturer: str | None = None
    """The manufacturer of the device e.g. 'Nest'."""

    sw_version: str | None = None
    """The firmware version string of the device e.g. '1.0.2'."""
