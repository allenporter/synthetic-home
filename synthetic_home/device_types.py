"""Data model for device type definitions.

These device devices and how they map to a set of entities. States can be set
on a device then translated into how that state affects entities.
"""

from functools import cache
from collections.abc import Generator
from dataclasses import dataclass, field
from importlib import resources
from importlib.resources.abc import Traversable
import logging
from typing import Any
import yaml

from mashumaro.codecs.yaml import yaml_decode
from mashumaro.exceptions import MissingField
from mashumaro import field_options, DataClassDictMixin
from mashumaro.types import SerializationStrategy

from .exceptions import SyntheticHomeError

__all__ = [
    "DeviceTypeRegistry",
    "DeviceType",
    "DeviceState",
    "EntityState",
    "EntityEntry",
    "load_device_type_registry",
]


_LOGGER = logging.getLogger(__name__)


DEVICE_TYPES_RESOURCE_PATH = resources.files().joinpath("registry")


class KeyedObjectListStrategy(SerializationStrategy):
    """A predefined entity state parser."""

    def __init__(self, object_strategy: SerializationStrategy) -> None:
        """Initialize KeyedObjectStrategy."""
        self._object_strategy = object_strategy

    def deserialize(self, value: Any) -> list[Any]:
        """Deserialize the object."""
        if not value:
            return []
        if isinstance(value, dict):
            return [
                self._object_strategy.deserialize((key, state))
                for key, state in value.items()
            ]
        raise ValueError(f"Expected 'dict' representing the object list, got: {value}")


@dataclass
class EntityEntry(DataClassDictMixin):
    """Defines an entity type.

    An entity is the lowest level object that maps to a behavior or trait of a device.
    """

    key: str
    """The entity description key"""

    attributes: dict[str, str | list[str]] = field(default_factory=dict)
    """Attributes supported by this entity."""


class EntityDictStrategy(SerializationStrategy):
    """A parser of the entity entry dict."""

    def deserialize(
        self, value: dict[str, dict[str, dict[str, str | list[str]]]]
    ) -> dict[str, list[EntityEntry]]:
        """Deserialize the object."""
        return {
            domain: [
                EntityEntry(key=key, attributes=attributes)
                for key, attributes in values.items()
            ]
            for domain, values in value.items()
        }


@dataclass
class EntityState(DataClassDictMixin):
    """Represents a pre-defined entity state."""

    domain: str
    """The domain of the entity."""

    key: str
    """The entity key identifying the entity."""

    state: str | dict[str, Any]
    """The values that make up the entity state."""

    @property
    def domain_key(self) -> str:
        """Return the full domain.key."""
        return f"{self.domain}.{self.key}"

    def merge(self, new: "EntityState") -> "EntityState":
        """Merge with an additional value."""
        merged_state: str | bool | dict[str, Any]
        if isinstance(new.state, dict):
            if isinstance(self.state, dict):
                merged_state = {
                    **self.state,
                    **new.state,
                }
            else:
                merged_state = {"state": self.state, **new.state}
        else:
            if isinstance(self.state, dict):
                merged_state = {
                    **self.state,
                    "state": new.state,
                }
            else:
                merged_state = new.state
        return EntityState(
            domain=self.domain,
            key=self.key,
            state=merged_state,
        )


def merge_entity_state_attributes(
    platform: str, entry: EntityEntry, entity_states: list[EntityState]
) -> EntityEntry:
    """Merge state values from an EntityEntry into a new EntityEntry."""
    for state in entity_states:
        if state.domain == platform and state.key == entry.key:
            if isinstance(state.state, dict):
                extras = state.state
            else:
                extras = {"state": state.state}
            return EntityEntry(
                key=entry.key,
                attributes={
                    **entry.attributes,
                    **extras,
                },
            )
    _LOGGER.debug("No state to merge: %s", entry)
    return entry


class EntityStateStrategy(SerializationStrategy):
    """A predefined entity state parser."""

    def deserialize(self, value: tuple[str, Any]) -> EntityState:
        """Deserialize the object."""
        key = value[0]
        state = value[1]
        parts = key.split(".")
        if len(parts) != 2:
            raise ValueError(f"Expected '<domain>.<entitiy-key>', got {key}")
        return EntityState(domain=parts[0], key=parts[1], state=state)


@dataclass
class DeviceState(DataClassDictMixin):
    """Represents a pre-defined device state.

    This is used to map a pre-defined devie state to the values that should be
    set on individual entities. These are typically "interesting" states of the
    device that can be enumerated during evaluation. For example, instead of
    explicitly specifying specific temperature values, a predefined state could
    implement "warm" and "cool".
    """

    name: str
    """An identifier that names this state."""

    entity_states: list[EntityState]
    """An identifier for this set of attributes used for labeling"""

    def merge(self, new_state: "DeviceState") -> "DeviceState":
        """Return a new DeviceState with merged entity states."""
        states = {entity.domain_key: entity for entity in self.entity_states}
        for entity in new_state.entity_states:
            if entity.domain_key in states:
                states[entity.domain_key] = states[entity.domain_key].merge(entity)
            else:
                states[entity.domain_key] = entity
        return DeviceState(name=self.name, entity_states=list(states.values()))


class DeviceStateStrategy(SerializationStrategy):
    """A predefined device state parser."""

    _child_strategy = KeyedObjectListStrategy(EntityStateStrategy())

    def deserialize(self, value: tuple[str, Any]) -> DeviceState:
        """Deserialize the object."""
        if not isinstance(value, tuple):
            raise ValueError(
                f"Expected 'tuple' representing the DeviceState object, got: {value}"
            )
        return DeviceState(
            name=value[0], entity_states=self._child_strategy.deserialize(value[1])
        )


@dataclass(frozen=True)
class DeviceType(DataClassDictMixin):
    """Defines of device type."""

    device_type: str
    """The identifier for the device e.g. 'smart-lock'"""

    desc: str
    """The human readable description of the device."""

    device_states: list[DeviceState] = field(
        metadata=field_options(
            serialization_strategy=KeyedObjectListStrategy(DeviceStateStrategy())
        ),
    )
    """A series of different attribute values that are most interesting to use during evaluation."""

    entities: dict[str, list[EntityEntry]] = field(
        metadata=field_options(serialization_strategy=EntityDictStrategy()),
        default_factory=dict,
    )
    """Entity platforms and their entity description keys"""

    @property
    def device_states_dict(self) -> dict[str, DeviceState]:
        """Get the predefined state by the specified key."""
        return {state.name: state for state in self.device_states}

    @property
    def entity_dict(self) -> dict[str, EntityEntry]:
        """Get a flat map of all entity entries."""
        return {
            f"{platform}.{entry.key}": entry
            for platform, entries in self.entities.items()
            for entry in entries
        }

    def __post_init__(self) -> None:
        """Validate the DeviceType."""
        entity_dict = self.entity_dict
        for device_state in self.device_states:
            for entity_state in device_state.entity_states:
                if entity_state.domain_key not in entity_dict:
                    raise ValueError(
                        f"Device '{self.device_type}' state '{device_state.name}' references "
                        f"invalid entity '{entity_state.domain_key}' not in {list(entity_dict.keys())}"
                    )


@dataclass
class DeviceTypeRegistry:
    """The registry of all DeviceType objects."""

    device_types: dict[str, DeviceType] = field(default_factory=dict)


def _read_device_types(
    device_types_path: Traversable,
) -> Generator[DeviceType, None, None]:
    """Read device types from the device type directory."""
    _LOGGER.debug("Loading device type registry from %s", device_types_path.absolute())  # type: ignore[attr-defined]

    for device_type_file in device_types_path.iterdir():
        if not device_type_file.name.endswith(".yaml"):
            continue
        try:
            with device_type_file.open("r") as f:
                content = f.read()
        except FileNotFoundError:
            raise SyntheticHomeError(
                f"Configuration file '{device_type_file}' does not exist"
            )

        try:
            device_type: DeviceType = yaml_decode(content, DeviceType)
        except MissingField as err:
            raise SyntheticHomeError(f"Unable to decode file {device_type_file}: {err}")
        except yaml.YAMLError as err:
            raise SyntheticHomeError(f"Unable to decode file {device_type_file}: {err}")
        if device_type_file.name != f"{device_type.device_type}.yaml":
            raise SyntheticHomeError(
                f"Device type '{device_type.device_type}' name does not match filename '{device_type_file.name}'"
            )

        yield device_type


@cache
def load_device_type_registry() -> DeviceTypeRegistry:
    """Load device types from the yaml configuration files."""
    device_types = {}
    for device_type in _read_device_types(DEVICE_TYPES_RESOURCE_PATH):
        if device_type.device_type in device_types:
            raise SyntheticHomeError(
                f"Device registry contains duplicate device type '{device_type.device_type}"
            )
        device_types[device_type.device_type] = device_type
    return DeviceTypeRegistry(device_types=device_types)
