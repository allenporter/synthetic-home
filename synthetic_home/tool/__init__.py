"""Synthetic home command line tools."""

__all__ = [
    "create_inventory",
    "export_inventory",
    "list_device_types",
]
