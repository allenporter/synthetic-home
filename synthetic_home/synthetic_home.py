"""Data model for home assistant synthetic home."""

from dataclasses import dataclass, field
import pathlib
import logging
import slugify
from typing import Any

from mashumaro.codecs.yaml import yaml_decode

from synthetic_home.exceptions import SyntheticHomeError
from synthetic_home.device_types import (
    DeviceTypeRegistry,
    DeviceStateStrategy,
    DeviceState,
    EntityEntry,
    merge_entity_state_attributes,
)
from . import common, inventory
from .inventory import DEFAULT_SEPARATOR
from .device_types import load_device_type_registry

__all__ = [
    "SyntheticHome",
    "Device",
    "build_device_state",
    "load_synthetic_home",
    "read_config_content",
]


_LOGGER = logging.getLogger(__name__)


@dataclass
class Device:
    """A synthetic device."""

    name: str
    """A human readable name for the device."""

    device_type: str | None = None
    """The type of the device in the device registry that determines how it maps to entities."""

    device_info: common.DeviceInfo | None = None
    """Device make and model information."""

    device_state: str | DeviceState | None = None
    """A list of pre-canned RestorableStateAttributes specified by the key.

    These are used for restoring a device into a specific state supported by the
    device type. This is used to use a label rather than specifying low level
    entity details. This is an alternative to specifying low level attributes above.

    Restorable attributes overwrite normal attributes since they can be reloaded
    at runtime.
    """

    entity_entries: dict[str, list[EntityEntry]] = field(default_factory=dict)
    """The validated set of entity entries"""

    def merge(
        self,
        device_state: DeviceState | None = None,
        entity_entries: dict[str, list[EntityEntry]] | None = None,
    ) -> "Device":
        """Merge the existing device with a new device state."""
        return Device(
            name=self.name,
            device_type=self.device_type,
            device_info=self.device_info,
            device_state=device_state or self.device_state,
            entity_entries=entity_entries or self.entity_entries,
        )


def build_device_state(device: Device, registry: DeviceTypeRegistry) -> Device:
    """Validate the device and return a new instance."""
    if (device_type := registry.device_types.get(device.device_type or "")) is None:
        raise SyntheticHomeError(
            f"Device {device} has device_type {device.device_type} not found in: {registry.device_types}"
        )

    if device.device_state is not None:
        # Lookup a device state and merge it into the current state
        if (
            isinstance(device.device_state, str)
            and device.device_state not in device_type.device_states_dict
        ):
            raise SyntheticHomeError(
                f"Device {device} has state '{device.device_state}' not in: {device_type.device_states_dict}"
            )
        if isinstance(device.device_state, dict):
            _LOGGER.debug(
                "Parsing device state from dictionary: %s", device.device_state
            )
            strategy = DeviceStateStrategy()
            device = device.merge(
                device_state=strategy.deserialize(("custom", device.device_state))
            )
            _LOGGER.debug("Parsed=%s", device)

    device_state: DeviceState | None = None
    _LOGGER.debug("Checking device state: %s", device.device_state)
    if (
        device.device_state is None or isinstance(device.device_state, DeviceState)
    ) and device_type.device_states:
        # Pick the first device state as the default
        device_state = device_type.device_states[0]
        if isinstance(device.device_state, DeviceState):
            device_state = device_state.merge(device.device_state)
    elif device.device_state is not None and isinstance(device.device_state, str):
        device_state = device_type.device_states_dict[device.device_state]
    else:
        raise SyntheticHomeError(f"Device did not declare a device state: {device}")

    _LOGGER.debug("Merging entity attributes for device state: %s", device_state)
    entity_entries = {
        platform: [
            merge_entity_state_attributes(
                platform, entity_entry, device_state.entity_states
            )
            for entity_entry in entity_entries
        ]
        for platform, entity_entries in device_type.entities.items()
    }
    return device.merge(device_state=device_state, entity_entries=entity_entries)


@dataclass
class SyntheticHome:
    """Data about a synthetic home."""

    # Devices by area
    devices: dict[str, list[Device]] = field(default_factory=dict)

    # Services for the home not for a specific area
    services: list[Device] = field(default_factory=list)

    # Device types supported by the home.
    device_type_registry: DeviceTypeRegistry | None = None

    def __post_init__(self) -> None:
        """Build the complete device state."""
        if self.device_type_registry is None:
            self.device_type_registry = load_device_type_registry()
        self.devices = {
            key: [
                build_device_state(device, self.device_type_registry)
                for device in devices
            ]
            for key, devices in self.devices.items()
        }
        self.services = [
            build_device_state(device, self.device_type_registry)
            for device in self.services
        ]


def read_config_content(config_file: pathlib.Path) -> str:
    """Create configuration file content, exposed for patching."""
    with config_file.open("r") as f:
        return f.read()


def load_synthetic_home(config_file: pathlib.Path) -> SyntheticHome:
    """Load synthetic home configuration from disk."""
    try:
        content = read_config_content(config_file)
    except FileNotFoundError:
        raise SyntheticHomeError(f"Configuration file '{config_file}' does not exist")
    try:
        return yaml_decode(content, SyntheticHome)
    except ValueError as err:
        raise SyntheticHomeError(f"Could not parse config file '{config_file}': {err}")


def yaml_state_value(v: Any) -> Any:
    """Convert a entity state value to yaml."""
    if isinstance(v, bool) or isinstance(v, float) or isinstance(v, list):
        return v
    return str(v)


def build_entities(area_id: str | None, device_entry: Device) -> list[inventory.Entity]:
    """Build the set of entities for the device entry."""
    entities = []
    device_name = device_entry.name.replace("_", " ").title()
    device_id = slugify.slugify(device_entry.name, separator=DEFAULT_SEPARATOR)
    for platform, entity_entries in device_entry.entity_entries.items():
        for entity_entry in entity_entries:
            # Each entity in this platform needs a unique name, but
            # if the key is in the name it's the primary to avoid "Motion motion"
            entity_name = device_name
            if platform == "sensor" or (
                platform == "binary_sensor"
                and entity_entry.key not in device_entry.name.lower()
            ):
                entity_name = f"{device_name} {entity_entry.key.capitalize()}"
            entity_id = f"{platform}.{slugify.slugify(entity_name, separator=DEFAULT_SEPARATOR)}"
            entity = inventory.Entity(
                name=entity_name,
                id=entity_id,
                device=device_id,
            )
            if area_id:
                entity.area = area_id
            attributes: dict[str, str | list[str] | int | float] = {}
            if entity_entry.attributes:
                attributes.update(entity_entry.attributes)
            state = attributes.pop("state", None)
            if state is not None:
                entity.state = yaml_state_value(state)
            if attributes:
                entity.attributes = attributes
            entities.append(entity)
    return entities


def build_inventory(home: SyntheticHome) -> inventory.Inventory:
    """Build a home inventory from the synthetic home definition.

    This is a flattened set of areas, entities, and devices.
    """

    inv = inventory.Inventory()
    pairs: list[tuple[str | None, list[Device]]] = [*home.devices.items()]
    if home.services:
        pairs.append((None, home.services))

    entities = []
    for area_name, devices in pairs:
        if area_name:
            area_id = slugify.slugify(area_name, separator=DEFAULT_SEPARATOR)
            inv.areas.append(inventory.Area(name=area_name, id=area_id))
        else:
            area_id = None

        for device_entry in devices:
            # Make computer generated device names more friendly
            device_name = device_entry.name.replace("_", " ").title()
            device_id = slugify.slugify(device_entry.name, separator=DEFAULT_SEPARATOR)
            device = inventory.Device(
                name=device_name,
                id=device_id,
                info=device_entry.device_info,
            )
            if area_id:
                device.area = area_id
            inv.devices.append(device)
            entities.extend(build_entities(area_id, device_entry))
    if entities:
        inv.entities = entities
    return inv
