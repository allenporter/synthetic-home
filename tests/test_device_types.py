"""Test for device_types."""

from synthetic_home import device_types


def test_load_device_type_registry() -> None:
    """Test loading the device type registry."""

    reg = device_types.load_device_type_registry()
    for name, device_type in reg.device_types.items():
        assert name
        assert device_type.device_type
        assert device_type.desc
        assert isinstance(device_type.device_states_dict, dict)
        assert device_type.entities, "Device must have at least one entity"
        for domain, entity_entries in device_type.entities.items():
            assert domain
            assert entity_entries, "Domain must have at least one entity entry"
            for entry in entity_entries:
                assert entry.key
                assert (
                    entry.attributes is not None
                ), f"Entity for device {name} has no attributes"


def test_camera_device() -> None:
    """Test the camera entry in the device registry."""

    reg = device_types.load_device_type_registry()
    assert reg.device_types
    assert "camera" in reg.device_types
    camera = reg.device_types["camera"]
    assert camera
    assert (
        camera.desc == "A video camera that supports motion detection and other events."
    )
    assert list(camera.device_states_dict.keys()) == [
        "idle",
        "motion-detected",
        "person-detected",
        "sound-detected",
    ]
    assert [
        (entry.key, entry.attributes)
        for entry in camera.entities.get("binary_sensor", [])
    ] == [
        (
            "motion",
            {"device_class": "binary_sensor.BinarySensorDeviceClass.MOTION"},
        ),
        (
            "person",
            {"device_class": "binary_sensor.BinarySensorDeviceClass.OCCUPANCY"},
        ),
        (
            "sound",
            {"device_class": "binary_sensor.BinarySensorDeviceClass.SOUND"},
        ),
    ]
