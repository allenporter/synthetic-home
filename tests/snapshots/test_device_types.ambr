# serializer version: 1
# name: test_device_state_to_entity_state[camera-idle]
  list([
    EntityState(domain='binary_sensor', key='motion', state=False),
    EntityState(domain='binary_sensor', key='person', state=False),
    EntityState(domain='binary_sensor', key='sound', state=False),
  ])
# ---
# name: test_device_state_to_entity_state[camera-motion-detected]
  list([
    EntityState(domain='binary_sensor', key='motion', state=True),
  ])
# ---
# name: test_device_state_to_entity_state[camera-person-detected]
  list([
    EntityState(domain='binary_sensor', key='person', state=True),
  ])
# ---
# name: test_device_state_to_entity_state[camera-sound-detected]
  list([
    EntityState(domain='binary_sensor', key='sound', state=True),
  ])
# ---
# name: test_device_state_to_entity_state[door-sensor-battery-low]
  list([
    EntityState(domain='binary_sensor', key='door', state=False),
    EntityState(domain='binary_sensor', key='battery', state=True),
    EntityState(domain='sensor', key='battery', state=10),
  ])
# ---
# name: test_device_state_to_entity_state[door-sensor-closed]
  list([
    EntityState(domain='binary_sensor', key='door', state=False),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[door-sensor-open]
  list([
    EntityState(domain='binary_sensor', key='door', state=True),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[exhaust-fan-humid]
  list([
    EntityState(domain='sensor', key='humidity', state=85),
  ])
# ---
# name: test_device_state_to_entity_state[exhaust-fan-off]
  list([
    EntityState(domain='fan', key='exhaust_fan', state={'is_on': False}),
    EntityState(domain='sensor', key='humidity', state=45),
  ])
# ---
# name: test_device_state_to_entity_state[exhaust-fan-on]
  list([
    EntityState(domain='fan', key='exhaust_fan', state={'is_on': True}),
    EntityState(domain='sensor', key='humidity', state=45),
  ])
# ---
# name: test_device_state_to_entity_state[fan-oscilating-off]
  list([
    EntityState(domain='fan', key='osciliating_fan', state={'is_on': False, 'oscillating': False}),
  ])
# ---
# name: test_device_state_to_entity_state[fan-oscilating-on]
  list([
    EntityState(domain='fan', key='osciliating_fan', state={'is_on': True, 'oscillating': False}),
  ])
# ---
# name: test_device_state_to_entity_state[fan-oscilating-oscillating]
  list([
    EntityState(domain='fan', key='osciliating_fan', state={'is_on': True, 'oscillating': True}),
  ])
# ---
# name: test_device_state_to_entity_state[garage-door-closed]
  list([
    EntityState(domain='cover', key='garage-door', state=False),
    EntityState(domain='light', key='light', state=False),
  ])
# ---
# name: test_device_state_to_entity_state[garage-door-open]
  list([
    EntityState(domain='cover', key='garage-door', state=True),
    EntityState(domain='light', key='light', state=True),
  ])
# ---
# name: test_device_state_to_entity_state[gate-closed]
  list([
    EntityState(domain='cover', key='gate', state=False),
  ])
# ---
# name: test_device_state_to_entity_state[gate-open]
  list([
    EntityState(domain='cover', key='gate', state=True),
  ])
# ---
# name: test_device_state_to_entity_state[heat-pump-heating]
  list([
    EntityState(domain='climate', key='heat-pump', state={'hvac_mode': 'heat', 'hvac_action': 'heating', 'current_temperature': 25, 'target_temperature': 27}),
    EntityState(domain='sensor', key='temperature', state=25),
  ])
# ---
# name: test_device_state_to_entity_state[heat-pump-off]
  list([
    EntityState(domain='climate', key='heat-pump', state={'hvac_mode': 'off', 'hvac_action': 'off', 'current_temperature': 22, 'target_temperature': 22}),
    EntityState(domain='sensor', key='temperature', state=22),
  ])
# ---
# name: test_device_state_to_entity_state[hvac-cooling]
  list([
    EntityState(domain='climate', key='hvac', state={'hvac_mode': 'cool', 'hvac_action': 'cooling', 'current_temperature': 25, 'target_temperature': 23}),
    EntityState(domain='sensor', key='temperature', state={'native_value': 25}),
    EntityState(domain='sensor', key='humidity', state={'native_value': 45}),
  ])
# ---
# name: test_device_state_to_entity_state[hvac-off]
  list([
    EntityState(domain='climate', key='hvac', state={'hvac_mode': 'off', 'hvac_action': 'off', 'current_temperature': 22, 'target_temperature': 22}),
    EntityState(domain='sensor', key='temperature', state={'native_value': 22}),
    EntityState(domain='sensor', key='humidity', state={'native_value': 45}),
  ])
# ---
# name: test_device_state_to_entity_state[hvac-very-low]
  list([
    EntityState(domain='climate', key='hvac', state={'hvac_mode': 'cool', 'hvac_action': 'cooling', 'current_temperature': 25, 'target_temperature': 12}),
    EntityState(domain='sensor', key='temperature', state={'native_value': 25}),
    EntityState(domain='sensor', key='humidity', state={'native_value': 45}),
  ])
# ---
# name: test_device_state_to_entity_state[light-dimmable-dim]
  list([
    EntityState(domain='light', key='light-dimmable', state={'brightness': 5}),
  ])
# ---
# name: test_device_state_to_entity_state[light-dimmable-off]
  list([
    EntityState(domain='light', key='light-dimmable', state={'brightness': 0}),
  ])
# ---
# name: test_device_state_to_entity_state[light-dimmable-on]
  list([
    EntityState(domain='light', key='light-dimmable', state={'brightness': 100}),
  ])
# ---
# name: test_device_state_to_entity_state[light-off]
  list([
    EntityState(domain='light', key='light', state='off'),
  ])
# ---
# name: test_device_state_to_entity_state[light-on]
  list([
    EntityState(domain='light', key='light', state='on'),
  ])
# ---
# name: test_device_state_to_entity_state[light-rgbw-off]
  list([
    EntityState(domain='light', key='light-rgbw', state={'rgbw_color': [0, 0, 0, 0]}),
  ])
# ---
# name: test_device_state_to_entity_state[light-rgbw-red]
  list([
    EntityState(domain='light', key='light-rgbw', state={'rgbw_color': [255, 0, 0, 0]}),
  ])
# ---
# name: test_device_state_to_entity_state[light-rgbw-white]
  list([
    EntityState(domain='light', key='light-rgbw', state={'rgbw_color': [0, 0, 0, 255]}),
  ])
# ---
# name: test_device_state_to_entity_state[lock-sensor-locked]
  list([
    EntityState(domain='binary_sensor', key='lock', state=False),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[lock-sensor-unlocked]
  list([
    EntityState(domain='binary_sensor', key='lock', state=True),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[motion-sensor-battery-low]
  list([
    EntityState(domain='binary_sensor', key='motion', state=False),
    EntityState(domain='binary_sensor', key='battery', state=True),
    EntityState(domain='sensor', key='battery', state=15),
  ])
# ---
# name: test_device_state_to_entity_state[motion-sensor-off]
  list([
    EntityState(domain='binary_sensor', key='motion', state=False),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[motion-sensor-on]
  list([
    EntityState(domain='binary_sensor', key='motion', state=True),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[sensor-zero]
  list([
    EntityState(domain='sensor', key='generic', state={'native_value': 0}),
  ])
# ---
# name: test_device_state_to_entity_state[smart-blinds-closed]
  list([
    EntityState(domain='cover', key='blind', state=False),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[smart-blinds-open]
  list([
    EntityState(domain='cover', key='blind', state=True),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[smart-lock-locked]
  list([
    EntityState(domain='lock', key='lock', state='locked'),
    EntityState(domain='binary_sensor', key='lock', state=False),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[smart-lock-unlocked]
  list([
    EntityState(domain='lock', key='lock', state='unlocked'),
    EntityState(domain='binary_sensor', key='lock', state=True),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[smart-plug-off]
  list([
    EntityState(domain='switch', key='outlet', state=False),
    EntityState(domain='sensor', key='energy', state=0),
  ])
# ---
# name: test_device_state_to_entity_state[smart-plug-on]
  list([
    EntityState(domain='switch', key='outlet', state=True),
    EntityState(domain='sensor', key='energy', state=1),
  ])
# ---
# name: test_device_state_to_entity_state[smart-speaker-off]
  list([
    EntityState(domain='media_player', key='speaker', state='off'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-speaker-paused]
  list([
    EntityState(domain='media_player', key='speaker', state='paused'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-speaker-playing]
  list([
    EntityState(domain='media_player', key='speaker', state='playing'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-sprinkler-closed]
  list([
    EntityState(domain='valve', key='sprinkler', state='closed'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-sprinkler-open]
  list([
    EntityState(domain='valve', key='sprinkler', state='open'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-tv-off]
  list([
    EntityState(domain='media_player', key='speaker', state='off'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-tv-paused]
  list([
    EntityState(domain='media_player', key='speaker', state='paused'),
  ])
# ---
# name: test_device_state_to_entity_state[smart-tv-playing]
  list([
    EntityState(domain='media_player', key='speaker', state='playing'),
  ])
# ---
# name: test_device_state_to_entity_state[switch-off]
  list([
    EntityState(domain='switch', key='switch', state=False),
  ])
# ---
# name: test_device_state_to_entity_state[switch-on]
  list([
    EntityState(domain='switch', key='switch', state=True),
  ])
# ---
# name: test_device_state_to_entity_state[temperature-sensor-cold]
  list([
    EntityState(domain='sensor', key='temperature', state=38),
    EntityState(domain='sensor', key='humidity', state=60),
  ])
# ---
# name: test_device_state_to_entity_state[temperature-sensor-hot]
  list([
    EntityState(domain='sensor', key='temperature', state=90),
    EntityState(domain='sensor', key='humidity', state=35),
  ])
# ---
# name: test_device_state_to_entity_state[temperature-sensor-mild]
  list([
    EntityState(domain='sensor', key='temperature', state=70),
    EntityState(domain='sensor', key='humidity', state=45),
  ])
# ---
# name: test_device_state_to_entity_state[todo-list-empty]
  list([
    EntityState(domain='todo', key='todo-list', state={'todo_items': []}),
  ])
# ---
# name: test_device_state_to_entity_state[vacuum-cleaning]
  list([
    EntityState(domain='vacuum', key='vacuum', state='cleaning'),
  ])
# ---
# name: test_device_state_to_entity_state[vacuum-off]
  list([
    EntityState(domain='vacuum', key='vacuum', state='docked'),
  ])
# ---
# name: test_device_state_to_entity_state[vacuum-paused]
  list([
    EntityState(domain='vacuum', key='vacuum', state='paused'),
  ])
# ---
# name: test_device_state_to_entity_state[vacuum-returning]
  list([
    EntityState(domain='vacuum', key='vacuum', state='returning'),
  ])
# ---
# name: test_device_state_to_entity_state[water-valve-closed]
  list([
    EntityState(domain='valve', key='water-valve', state={'current_valve_position': 0}),
    EntityState(domain='sensor', key='meter-reading', state=2000),
    EntityState(domain='sensor', key='battery', state=75),
  ])
# ---
# name: test_device_state_to_entity_state[water-valve-open]
  list([
    EntityState(domain='valve', key='water-valve', state={'current_valve_position': 100}),
    EntityState(domain='sensor', key='meter-reading', state=2001),
    EntityState(domain='sensor', key='battery', state=75),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-cloudy]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'cloudy', 'native_temperature': 60, 'humidity': 70, 'native_wind_speed': 10}),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-cold]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'partlycloudy', 'native_temperature': 20, 'humidity': 60, 'native_wind_speed': 5}),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-fair]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'sunny', 'native_temperature': 72, 'humidity': 60, 'native_wind_speed': 5}),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-hot]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'sunny', 'native_temperature': 90, 'humidity': 40, 'native_wind_speed': 10}),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-rainy]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'rainy', 'native_temperature': 50, 'humidity': 80, 'native_wind_speed': 15}),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-snowy]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'snowy', 'native_temperature': 32, 'humidity': 90, 'native_wind_speed': 20}),
  ])
# ---
# name: test_device_state_to_entity_state[weather-service-windy]
  list([
    EntityState(domain='weather', key='weather', state={'condition': 'windy', 'native_temperature': 65, 'humidity': 50, 'native_wind_speed': 30}),
  ])
# ---
# name: test_device_state_to_entity_state[window-sensor-battery-low]
  list([
    EntityState(domain='binary_sensor', key='window', state=False),
    EntityState(domain='binary_sensor', key='battery', state=True),
    EntityState(domain='sensor', key='battery', state=10),
  ])
# ---
# name: test_device_state_to_entity_state[window-sensor-closed]
  list([
    EntityState(domain='binary_sensor', key='window', state=False),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
# name: test_device_state_to_entity_state[window-sensor-open]
  list([
    EntityState(domain='binary_sensor', key='window', state=True),
    EntityState(domain='sensor', key='battery', state=90),
  ])
# ---
